src/main/java/org/testng/internal/FactoryMethod.java

```diff
@@ -82,8 +82,14 @@ private void init(Object instance, IAnnotationFinder annotationFinder, Construct
     Utils.checkReturnType(com.getMethod(), Object[].class, IInstanceInfo[].class);
     Class<?> declaringClass = com.getDeclaringClass();
     if (instance != null && !declaringClass.isAssignableFrom(instance.getClass())) {
-      throw new TestNGException(
-          "Mismatch between instance/method classes:" + instance.getClass() + " " + declaringClass);
+      Class<?> cls = instance.getClass();
```

Comment:
I think the following is a bit more readable:
```java
      if (instance instanceof IParameterInfo) {
        instance = ((IParameterInfo) instance).getInstance();
      }
      Class<?> cls = instance.getClass();
```

Reply:
Fixed.